File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.java
class Solution {
    public int firstStableIndex(int[] nums, int k) {
        if (nums.length == 0) return 0;

        int n = nums.length;
        int instability;

        // Prefix maximum
        int[] leftMax = new int[n];
        int currMax = nums[0];

        for (int i = 0; i < n; i++) {
            currMax = Math.max(currMax, nums[i]);
            leftMax[i] = currMax;
        }

        // Suffix minimum
        int[] rightMin = new int[n];
        int currMin = nums[n - 1];

        for (int i = n - 1; i >= 0; i--) {
            currMin = Math.min(currMin, nums[i]);
            rightMin[i] = currMin;
        }

        // Check instability
        for (int i = 0; i < n; i++) {
            instability = leftMax[i] - rightMin[i];

            if (instability <= k) {
                return i;
            }
        }
        return -1;
    }
}

File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        n = len(nums)

        left_max = []
        curr_max = nums[0]

        # Find for left MAX
        for i in range(n):
            curr_max = max(curr_max, nums[i])
            left_max.append(curr_max)

        right_min = [0] * n
        curr_min = nums[-1]

        # Find for Right MIN
        for i in range(n - 1, -1, -1):
            curr_min = min(curr_min, nums[i])
            right_min[i] = curr_min

        # FInd stable index
        for i in range(n):
            instability = left_max[i] - right_min[i]

            if instability <= k:
                return i

        return -1